src/shogun/preprocessor/FisherLDA.cpp

```diff
@@ -109,175 +113,207 @@ bool CFisherLDA::fit(CFeatures *features, CLabels *labels, int32_t num_dimension
+		linalg::add_col_vec(
+		    feature_centered, i, mean_class[labels_vector[i]], feature_centered,
+		    (float64_t)1, (float64_t)-1);
 
 	if ((m_method==CANVAR_FLDA) ||
 			(m_method==AUTO_FLDA && num_vectors<num_features))
 	{
-		// holds the  fmatrix for each class
-		vector<MatrixXd> centered_class_i(C);
-		VectorXd temp=num_class;
-		MatrixXd Sw=MatrixXd::Zero(num_features, num_features);
-		for (i=0; i<C; i++)
+		// holds the feature matrix for each class
+		vector<SGMatrix<float64_t>> centered_class(C);
+		vector<index_t> centered_class_col(C);
+
+		SGMatrix<float64_t> Sw(num_features, num_features);
+		linalg::zero(Sw);
+		for (index_t i = 0; i < C; ++i)
 		{
-			centered_class_i[i]=MatrixXd::Zero(num_features, num_class[i]);
-			for (j=0; j<num_vectors; j++)
-				if (i==lvector[j])
-					centered_class_i[i].col(num_class[i]-(temp[i]--))
-						=fmatrix.col(j);
-			Sw+=(centered_class_i[i]*centered_class_i[i].transpose())
-				*num_class[i]/(float64_t)(num_class[i]-1);
+			centered_class[i] = SGMatrix<float64_t>(num_features, num_class[i]);
+			linalg::zero(centered_class[i]);
+		}
+		for (index_t i = 0; i < num_vectors; i++)
+		{
+			index_t c = (index_t)labels_vector[i];
+			// TODO: SGMatrix.set_column(i, v)
+			SGVector<float64_t> v(
+			    feature_centered.get_column_vector(i), num_features, false);
+			linalg::add_col_vec(
+			    centered_class[c], centered_class_col[c], v, centered_class[c]);
+			centered_class_col[c]++;
+		}
+		for (index_t i = 0; i < C; ++i)
+		{
+			auto tmp = linalg::matrix_prod(
+			    centered_class[i], centered_class[i], false, true);
+			linalg::scale(
+			    tmp, tmp, num_class[i] / (float64_t)(num_class[i] - 1));
+			linalg::add(Sw, tmp, Sw);
 		}
 
-		// within class matrix for cannonical variates implementation
-		MatrixXd Sb(num_features, C);
-		for (i=0; i<C; i++)
-		Sb.col(i)=sqrt(num_class[i])*(mean_class[i]-mean_total);
-
-		MatrixXd fmatrix1=Map<MatrixXd>(feature_matrix.matrix, num_features,
-									num_vectors);
+		// regularization
+		float64_t trace = linalg::trace(Sw);
+		SGMatrix<float64_t> id(Sw.num_rows, Sw.num_rows);
+		linalg::identity(id);
+		linalg::add(
+		    Sw, id, Sw, (1.0 - m_gamma), trace * (m_gamma) / num_features);
+
+		// within class matrix for canonical variates implementation
+		SGMatrix<float64_t> Sb(num_features, C);
+		linalg::zero(Sb);
+		for (index_t i = 0; i < C; i++)
+		{
+			// TODO: SGMatrix.get_column(i) -> SGVector
+			SGVector<float64_t> col(
+			    Sb.get_column_vector(i), num_features, false);
+			linalg::add(mean_class[i], m_mean_vector, col, 1.0, -1.0);
+			linalg::scale(col, col, sqrt(num_class[i]));
+		}
 
-		JacobiSVD<MatrixXd> svd(fmatrix1, ComputeThinU | ComputeThinV);
+		index_t r = CMath::min(num_vectors, num_features);
+		SGMatrix<float64_t> U(num_features, r);
+		SGVector<float64_t> singularValues(r);
+		linalg::svd(feature_matrix, singularValues, U);
 		// basis to represent the solution
-		MatrixXd Q;
+		SGMatrix<float64_t> Q;
```

Comment:
is this design fixed in the way that no matter what the input features are, Q is float64? I am ok, just asking